clang/lib/Sema/SemaChecking.cpp

```diff
@@ -13861,6 +13897,36 @@ Sema::CheckReturnValExpr(Expr *RetValExp, QualType lhsType,
     CheckPPCMMAType(RetValExp->getType(), ReturnLoc);
 }
 
+/// Diagnose comparison to NAN or INFINITY in fast math modes.
+/// The comparison to NaN or INFINITY is always false in
+/// fast modes: float evaluation will not result in inf or nan.
+void Sema::CheckInfNaNFloatComparison(SourceLocation Loc, Expr *LHS, Expr *RHS,
```

Comment:
The basic idea I have is that it's UB to generate a constant infinity or NaN if infinity/NaN is disabled. Instead of performing special logic for float comparisons, I think the logic should live in `Sema::ActOnNumericConstant()` so that we capture any time a float or infinity constant is used. e.g. `inf + 1` should also diagnose.

Then the tautological comparison component to this could be implemented by looking for calls to `__builtin_isinf` or `__builtin_isnan` and claim those are tautologically false (as a follow-up patch). WDYT?

Reply:
I have actually tried that but I noticed that `getSpellingOfSingleCharacterNumericConstant` of inf/nan will both return "0" so I will have no way to distinguish between inf + 1 and 0 + 1 and both will generate a warning. Unless there is other way of finding out that the token is nan/inf?

Reply:
Why are we on that path in the first place? The token length shouldn't be one character long for either `inf` or `nan` right? I was thinking you should be able to call `NumericLiteralParser::GetFloatValue()` within the `} else if (Literal.isFloatingLiteral()) {` block to get an `APFloat` that you can call `isInfinity()` or `isNaN()` on and then diagnose at that point.

Reply:
@AaronBallman Thanks.